src/Control/Concurrent/TOrderedQueue.hs

```diff
@@ -0,0 +1,106 @@
+-- |
+-- Module      :  Control.Concurrent.TOrderedQueue
+-- Copyright   :  (c) Arista Networks, 2022-2023
+-- License     :  Apache License 2.0, see COPYING
+--
+-- Stability   :  stable
+-- Portability :  non-portable (GHC extensions)
+--
+-- Definitions for the 'TOrderedQueue' container.
+--
+-- @since 1.0.0
+module Control.Concurrent.TOrderedQueue where
+
+--------------------------------------------------------------------------------
+
+import Relude
+
+import Control.Concurrent.STM
+  ( TQueue,
+    newTQueue,
+    newTQueueIO,
+    retry,
+    tryReadTQueue,
+    writeTQueue,
+  )
+
+import Data.PQueue.Prio.Min (MinPQueue, insert, minViewWithKey)
+
+data SequenceId
+  = Unordered
+  | SequenceId !Natural
+  deriving stock (Eq, Ord, Show)
+
+-- | A value `a` annotated with a `SequenceId`
+data Sequenced a = Sequenced
+  { idx :: !SequenceId
+  , val :: !a
+  }
+  deriving stock (Show)
+
+instance Eq (Sequenced a) where
+  (==) = (==) `on` idx
+
+instance Ord (Sequenced a) where
+  (<=) = (<=) `on` idx
+
+-- | A queue in which messages written to the queue can be annotated with a `SequenceId`, and messages read from
+-- the queue will always be seen in order by `SequenceId`, even if the messages are written out-of-order.
+-- Messages written to the queue annotated with `Unordered` will be read before any ordered messages and will be
+-- processed in FIFO order.
+data TOrderedQueue a
+  = TOrderedQueue
+      !(TVar Natural) -- Sequence counter indicating what the sequence id of the next message read should be
+      !(TVar (MinPQueue Natural a)) -- A min heap for holding messages annotated by `SequenceId`
+      !(TQueue a) -- A regular TQueue used for processing `Unordered` messages
```

Comment:
So this is here essentially as a fallback? To be backwards compatible with protocol v3 with the same code?

Is there ever a case where both queues can fill up?

Reply:
It is primarily for backwards compatibility. It could also come into play if an out-of-band process wanted to throw an error into a session without having access to the current sequence id. (There's nothing in here that does that currently, but there are a few places where errors are currently discarded and this could potentially be a way to address that in the future)

I can't think of any situations in normal use where both queues would be in heavy use simultaneously. But even if they were, I think it should still all work fine. Do you see a potential issue here? 

(Also, the queues are unbounded, so there should never be a situation where either is full.)

Reply:
I don't see an issue. I was just curious because the two queue approach didn't make a lot of sense just reading through. I'd love something similar to your explanation in the code.